cloud-init/qemu_performance.py

```diff
@@ -0,0 +1,555 @@
+ppa:<custom_ppa>.
+
+Launch multiple instances up to --number-of-launches of control and upgraded
+cloudimages.
+
+Persist all metrics artifacts in --data-dir as JSON files.
+Calculate averages across all control runs and upgraded image samples.
+
+Highlight deltas between control verus upgraded averages which
+are greater than 0.1 seconds different and 20 percent different.
+
+
+REQUIREMENTS: 
+- sudo permissions to mount ISO images
+- mount-image-callback utility from cloud-image-utils deb package
+"""
+
+import json
+import logging
+import os
+import re
+import shutil
+import statistics
+import subprocess
+import sys
+import time
+from argparse import ArgumentParser, FileType
+from copy import deepcopy
+from pathlib import Path
+
+import pycloudlib
+
+# Number of original and upgraded boots to perform and evaluate
+BOOT_SAMPLES = 3
+
+# Do not report performance deltas where average delta for service between
+# control and new images is less than this number of seconds
+MIN_AVG_DELTA = 0.1
+
+# Do not report performance deltas for services where average delta between
+# control and new is below this percentage
+MIN_AVG_PERCENT_DELTA = 20
+
+
+MANDATORY_REPORT_KEYS = ("time_cloudinit_total", "time_systemd_userspace", "client_time_to_ssh")
+
+DEFAULT_PPA = "ppa:cloud-init-dev/daily"
+
+
+def retry_cmd(instance, cmd):
+    for retry_sleep in [0.25] * 400:
+        try:
+            return instance.execute(cmd)
+        except Exception:
+            time.sleep(retry_sleep)
+
+
+def update_cloud_init_in_img(img_path: str, deb_path: str, suffix=".modified"):
+    """Use mount-image-callback to install a known deb into an image"""
+    new_img_path = os.path.basename(img_path.replace(".img", f".img{suffix}"))
+    new_img_path = f"{os.getcwd()}/{new_img_path}"
+    shutil.copy(img_path, new_img_path)
+    subprocess.run(["sync"])
+    if deb_path.endswith(".deb"):
+        commands = [
+            [
+                "sudo",
+                "mount-image-callback",
+                "--system-mounts", "--system-resolvconf",
+                new_img_path,
+                "--",
+                "sh",
+                "-c",
+                f"chroot ${{MOUNTPOINT}} apt-get update -y; chroot ${{MOUNTPOINT}} apt-get install dhcpcd5 -y",
+            ],
+            [
+                "sudo",
+                "mount-image-callback",
+                new_img_path,
+                "--",
+                "sh",
+                "-c",
+                f"cp {deb_path} ${{MOUNTPOINT}}/.; chroot ${{MOUNTPOINT}} dpkg -i /{os.path.basename(deb_path)}",
+            ],
+        ]
+    elif deb_path.startswith("ppa:"):
+        commands = [
+            [
+                "sudo",
+                "mount-image-callback",
+                new_img_path,
+                "--",
+                "sh",
+                "-c",
+                f"add-apt-repository ppa:cloud-init-dev/daily -y; DEBIAN_FRONTEND=noninteractive apt-get install cloud-init -y",
```

Comment:
You just upgraded cloud-init on my local machine :smile: 

We need the chroot MOUNTPOINT here once we get it working.

Reply:
Oops, I've been providing the deb directly on the command line and hadn't been exercising this path.